File: frontend/src/services/featureExtractor.js
/**
 * CORRECT Feature Extraction for GREYC-NISLAB Dataset
 * 
 * This extractor properly handles:
 * - Multiple key presses (same key repeated)
 * - Backspace corrections
 * - Event order validation
 * - Proper pairing of keydown/keyup
 * 
 * Time scale: 100ns ticks (multiply ms by 10,000)
 */
const TIME_SCALE = 10000  // Convert milliseconds to 100ns ticks

export const extractFeatures = (keystrokeData) => {
  console.log('🔧 Starting feature extraction...')
  
  // 1. Sort events by timestamp
  const events = [...keystrokeData].sort((a, b) => a.timestamp - b.timestamp)
  
  if (events.length === 0) {
    console.warn('No events provided')
    return new Array(92).fill(0)
  }

  // 2. Filter out non-character keys AND Backspace
  // Only keep: letters, space, and punctuation that are part of the password
  const validKeys = new Set([
    'a','b','c','d','e','f','g','h','i','j','k','l','m',
    'n','o','p','q','r','s','t','u','v','w','x','y','z',
    ' '
  ])
  
  const filteredEvents = events.filter(event => {
    // Skip Backspace entirely (we'll handle corrections differently)
    if (event.key === 'Backspace') return false
    // Only keep character keys
    return validKeys.has(event.key)
  })
  
  console.log(`📊 Filtered to ${filteredEvents.length} events`)

  if (filteredEvents.length === 0) {
    console.warn('No valid events after filtering')
    return new Array(92).fill(0)
  }

  // 3. Properly pair keydown with keyup using queues
  // This handles repeated keys correctly (e.g., 't' appears multiple times)
  const keyQueues = {}
  const pairedEvents = []
  
  for (const event of filteredEvents) {
    if (event.type === 'keydown') {
      // Add to queue for this key
      if (!keyQueues[event.key]) {
        keyQueues[event.key] = []
      }
      keyQueues[event.key].push({
        timestamp: event.timestamp,
        key: event.key
      })
    } else if (event.type === 'keyup') {
      // Find the matching keydown (FIFO)
      if (keyQueues[event.key] && keyQueues[event.key].length > 0) {
        const downEvent = keyQueues[event.key].shift()
        // Only pair if up timestamp is AFTER down timestamp
        if (event.timestamp > downEvent.timestamp) {
          pairedEvents.push({
            key: event.key,
            downTime: downEvent.timestamp,
            upTime: event.timestamp
          })
        } else {
          console.warn('⚠️ Keyup before keydown, skipping:', event.key)
        }
      }
    }
  }

  console.log(`📊 Paired ${pairedEvents.length} key events`)

  if (pairedEvents.length < 2) {
    console.warn('Not enough paired events')
    return new Array(92).fill(0)
  }

  // 4. Verify event order (should be strictly increasing)
  let isValid = true
  for (let i = 1; i < pairedEvents.length; i++) {
    if (pairedEvents[i].downTime <= pairedEvents[i-1].downTime) {
      console.warn('⚠️ Events out of order at index', i)
      isValid = false
    }
  }

  if (!isValid) {
    console.warn('⚠️ Event order invalid - using subset')
    // Sort by downTime to fix ordering
    pairedEvents.sort((a, b) => a.downTime - b.downTime)
  }

  // 5. Extract press and release times (in 100ns ticks)
  const pressTimes = pairedEvents.map(e => e.downTime * TIME_SCALE)
  const releaseTimes = pairedEvents.map(e => e.upTime * TIME_SCALE)

  console.log(`📊 Press times: ${pressTimes.length}, Release times: ${releaseTimes.length}`)

  // 6. Build features
  const features = []

  // PP: Press-to-Press (time between consecutive keydowns)
  for (let i = 0; i < pressTimes.length - 1; i++) {
    const pp = pressTimes[i + 1] - pressTimes[i]
    // Validate: PP should be positive and reasonable (0-10,000,000 = 0-1000ms)
    if (pp > 0 && pp < 10000000) {
      features.push(pp)
    } else {
      console.warn(`⚠️ Invalid PP at index ${i}: ${pp}`)
      features.push(0)
    }
  }

  // RR: Release-to-Release (time between consecutive keyups)
  for (let i = 0; i < releaseTimes.length - 1; i++) {
    const rr = releaseTimes[i + 1] - releaseTimes[i]
    if (rr > 0 && rr < 10000000) {
      features.push(rr)
    } else {
      console.warn(`⚠️ Invalid RR at index ${i}: ${rr}`)
      features.push(0)
    }
  }

  // PR: Press-to-Release (hold duration for each key)
  for (let i = 0; i < pressTimes.length; i++) {
    const pr = releaseTimes[i] - pressTimes[i]
    if (pr > 0 && pr < 10000000) {
      features.push(pr)
    } else {
      console.warn(`⚠️ Invalid PR at index ${i}: ${pr}`)
      features.push(0)
    }
  }

  // RP: Release-to-Press (time between keyup and next keydown)
  for (let i = 0; i < pressTimes.length - 1; i++) {
    const rp = pressTimes[i + 1] - releaseTimes[i]
    if (rp > 0 && rp < 10000000) {
      features.push(rp)
    } else {
      console.warn(`⚠️ Invalid RP at index ${i}: ${rp}`)
      features.push(0)
    }
  }

  // 7. Pad or truncate to exactly 92 features
  while (features.length < 92) {
    features.push(0)
  }
  
  const result = features.slice(0, 92)

  // 8. Log statistics
  const nonZero = result.filter(f => f !== 0).length
  const negatives = result.filter(f => f < 0).length
  
  console.log(`📊 Extracted ${result.length} features`)
  console.log(`📊 Non-zero features: ${nonZero}`)
  console.log(`📊 Negative features: ${negatives}`)
  
  if (negatives > 0) {
    console.warn(`⚠️ Found ${negatives} negative features!`)
  }
  
  console.log(`📊 First 10 features:`, result.slice(0, 10).map(f => Math.round(f)))
  console.log(`📊 Last 10 features:`, result.slice(82, 92).map(f => Math.round(f)))

  return result
}

File: backend/services/feature_extractor.py
"""
Feature extraction from raw keystroke events.
Converts keydown/keyup timestamps to the 92-dimensional feature vector.
"""
from typing import List, Dict, Any
import numpy as np
from utils.logger import logger

class KeystrokeFeatureExtractor:
    """
    Extract features from keystroke events matching the GREYC-NISLAB dataset format.
    
    Features:
    - Press-to-Press (PP) times
    - Release-to-Release (RR) times  
    - Press-to-Release (PR) times (hold durations)
    - Release-to-Press (RP) times
    """
    
    def __init__(self):
        self.feature_count = 92
    
    def extract_features(self, keystroke_events: List[Dict[str, Any]]) -> List[float]:
        """
        Extract 92 features from raw keystroke events.
        
        Args:
            keystroke_events: List of events with 'key', 'type' (down/up), 'timestamp'
            
        Returns:
            List of 92 float features
        """
        try:
            if not keystroke_events:
                logger.warning("No keystroke events provided")
                return [0.0] * self.feature_count
            
            logger.info(f"Processing {len(keystroke_events)} events")
            
            # Sort events by timestamp
            events = sorted(keystroke_events, key=lambda x: x['timestamp'])
            logger.info(f"Sorted {len(events)} events by timestamp")
            
            # Group by key
            key_events = {}
            for event in events:
                key = event.get('key', '')
                event_type = event.get('type', '')
                
                if not key or event_type not in ['keydown', 'keyup']:
                    logger.warning(f"Skipping invalid event: {event}")
                    continue
                
                if key not in key_events:
                    key_events[key] = {'down': [], 'up': []}
                
                if event_type == 'keydown':
                    key_events[key]['down'].append(event['timestamp'])
                elif event_type == 'keyup':
                    key_events[key]['up'].append(event['timestamp'])
            
            logger.info(f"Grouped events into {len(key_events)} keys")
            
            # Calculate features
            features = []
            
            # Get list of keys in order of first appearance
            ordered_keys = []
            seen = set()
            for event in events:
                key = event.get('key', '')
                if key and key not in seen:
                    seen.add(key)
                    ordered_keys.append(key)
            
            logger.info(f"Processing keys in order: {ordered_keys[:10]}...")  # Show first 10 keys
            
            # For each key, compute features
            for i, key in enumerate(ordered_keys):
                events_for_key = key_events.get(key, {'down': [], 'up': []})
                down_times = events_for_key['down']
                up_times = events_for_key['up']
                
                # Need at least one down/up pair
                n_pairs = min(len(down_times), len(up_times))
                
                for j in range(n_pairs):
                    down_time = down_times[j]
                    up_time = up_times[j]
                    
                    # Press-to-Release (PR) - hold time
                    features.append(up_time - down_time)
                    
                    # Press-to-Press (PP) - time between consecutive presses
                    if j > 0 and j < len(down_times):
                        features.append(down_time - down_times[j-1])
                    
                    # Release-to-Release (RR) - time between consecutive releases
                    if j > 0 and j < len(up_times):
                        features.append(up_time - up_times[j-1])
                    
                    # Release-to-Press (RP) - time between release and next press
                    if j < n_pairs - 1 and j+1 < len(down_times):
                        features.append(down_times[j+1] - up_time)
            
            logger.info(f"Extracted {len(features)} raw features")
            
            # Pad or truncate to exactly 92 features
            while len(features) < self.feature_count:
                features.append(0.0)
            
            result = features[:self.feature_count]
            logger.info(f"Final feature vector length: {len(result)}")
            
            return result
            
        except Exception as e:
            logger.error(f"Error extracting features: {str(e)}")
            # Return zeros as fallback
            return [0.0] * self.feature_count